File: Hash Table/Minimum Window Substring/solve.cpp
class Solution {
/*
 * Time Complexity: O(N+K) N is length of string 's' and 't' respectively.
 * Space Complexity: O(N+K) same with above.
 * Problem statement here: https://leetcode.com/problems/minimum-window-substring/
 * The explaination can also be found on the above URL.
 */
public:
    string minWindow(string s, string t) {
       int freq[256] = {};
        
        int count = t.size();
        for (int i=0; i<t.size(); ++i) {
            ++freq[t[i]];
        }

    int minLen = s.length() + 1;
    int minStart = -1;
    int seen = 0;

    for (int start = 0, end = 0; end < s.size();) {
        if (--freq[s[end++]] >= 0) {
            seen++;
        }  

        if (seen >= count) {
	        while(++freq[s[start++]] <= 0) {};
            seen--;                
            if (minLen > end - start + 1) {
                minLen = end - start + 1;
                minStart = start - 1;
            }
        }
    }

    return minStart >= 0 ? s.substr(minStart, minLen) : "";
    }
};

File: Monotoic Queue/Maximum Circular Subarray/solve.java
class Solution {
/*
 * Time Complexity: O(N) N is length of input array
 * Space Complexity: O(1) same usage of space no matter how big the input is
 * Problem statement here: https://leetcode.com/problems/maximum-sum-circular-subarray/
 * The explaination can also be found on the above URL.
 */
    static int[] maxSubArray(int[] arr,int j) {
        int TotalSum = arr[j];
        int n = arr.length;
        int MaxSum = arr[j];
        int currentSum = arr[j];
        for (int idx = j + 1; idx < n + j; idx++) {
            int k = idx % n;
            int val1 = arr[k];
            TotalSum += val1;
            currentSum = Math.max(currentSum + val1, val1); // from kadane
            MaxSum = Math.max(MaxSum, currentSum);
        }
        int[] resArr = {MaxSum,TotalSum};
        return resArr;
    }
    static int minSubArray(int[] arr,int j) {
        int n = arr.length;
        int MinSum = arr[j];
        int currentSum = arr[j];
        for (int idx = j + 1; idx < n + j; idx++) {
            int k = idx % n;
            int val1 = arr[k];
            currentSum = Math.min(currentSum + val1, val1); // from kadane
            MinSum = Math.min(MinSum, currentSum);
        }
        return MinSum;
    }
    static public int maxSubarraySumCircular(int[] nums) {
        int n = nums.length;
        int CircularSum = nums[0];
        int[] resArr = maxSubArray(nums,0);
        int MaxNormaArrSum  = resArr[0];
        int TotalSum = resArr[1];
        int MinSum = minSubArray(nums,0);
        int Onlycircularsum = TotalSum - MinSum;
        if (MaxNormaArrSum<0){
            return MaxNormaArrSum;
        } else {
            return Math.max(Onlycircularsum, MaxNormaArrSum);
        }
    }
}